src/block/Beacon.php

```diff
@@ -32,9 +32,56 @@
 use pocketmine\item\Item;
 use pocketmine\math\Vector3;
 use pocketmine\player\Player;
-use function in_array;
 
 final class Beacon extends Transparent{
+	private const ALLOWED_BLOCK_IDS = [
+		BlockLegacyIds::IRON_BLOCK => true,
+		BlockLegacyIds::GOLD_BLOCK => true,
+		BlockLegacyIds::DIAMOND_BLOCK => true,
+		BlockLegacyIds::EMERALD_BLOCK => true
+		//TODO netherite block
+	];
+
+	private int $primaryEffect;
+	private int $secondaryEffect;
+
+	public function readStateFromWorld() : void{
+		parent::readStateFromWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileBeacon){
+			$this->primaryEffect = $tile->getPrimaryEffect();
+			$this->secondaryEffect = $tile->getSecondaryEffect();
+		}
+	}
+
+	public function writeStateToWorld() : void{
+		parent::writeStateToWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileBeacon){
+			$tile->setPrimaryEffect($this->primaryEffect);
+			$tile->setSecondaryEffect($this->secondaryEffect);
+		}
+	}
+
+	public function getPrimaryEffect() : int{
+		return $this->primaryEffect;
+	}
+
+	/** @return $this */
+	public function setPrimaryEffect(int $primaryEffect) : self{
+		$this->primaryEffect = $primaryEffect;
```

Comment:
Perhaps it would be necessary to check the effects sent, as stated in the channel?